### PR TITLE
Add new purchase price field to allow net and gross purchase prices

### PR DESCRIPTION
**Pull request by hanneswernery**
### 1. Why is this change necessary?

Currently the product and the payload of the order line items have a purchase price field of type `float`. Many users want to manage their purchase prices as _net_ prices. Some want to use different currencies when purchasing products.
Hence the more complex `PriceField` with gross and net prices as well as currencies would be more suitable.

### 2. What does this change do, exactly?

- A new purchase price field `purchase_prices` of type `PriceField` is added to the product definition as well as to the order line items (and some corresponding classes).
- The old `float` purchase price field (and some corresponding functions) is marked as deprecated.
- A migration is added to migrate the deprecated purchase price field to the new one.
- An ApiConverter is added to provide the old purchase price field for the api v1 and the new field for the api v2.
- A database trigger is added to keep the deprecated purchase price field in sync with the new one (updating either will update the other).
- The existing `LineItemPurchasePriceRule` is removed in favour of `LineItemPurchasePriceGrossRule` and `LineItemPurchasePriceNetRule`.
- Existing tests are updated and some new tests are added.
- The administration product view is updated to display the new purchase price field. The layout is open for discussion:

Screenshot 2020-02-27 at 11 10 06

### 4. Please link to the relevant issues (if any).

- https://issues.shopware.com/issues/NEXT-7666
- https://issues.shopware.com/issues/NEXT-5021

### 5. Checklist

- [x] I have written tests and verified that they fail without my change
- [x] I have squashed any insignificant commits
- [x] I have written or adjusted the documentation according to my changes
- [x] This change has comments for package types, values, functions, and non-obvious lines of code
- [x] I have read the contribution requirements and fulfil them.

### CODE REVIEW

File: src/Core/Checkout/Cart/Order/Transformer/PriceTransformer.php
Comment (shyim) [inline src/Core/Checkout/Cart/Order/Transformer/PriceTransformer.php]: Any specific reason why static? This cannot be extended by an Plugin

Comment (hanneswernery) [reply src/Core/Checkout/Cart/Order/Transformer/PriceTransformer.php]: The transformer methods are not part of any `PriceTransformer` instance. Hence they can and should be static.
I also oriented myself to the existing `AddressTransformer` https://github.com/shopware/platform/blob/b4124f2d5a04846af4e2dc643f43fdaf5e3dbbea/src/Core/Checkout/Cart/Order/Transformer/AddressTransformer.php

File: src/Core/Checkout/Cart/Rule/LineItemPurchasePriceGrossRule.php
Comment (OliverSkroblin) [inline src/Core/Checkout/Cart/Rule/LineItemPurchasePriceGrossRule.php]: Please merge the gross and net rule to a single pruchase price rule with a parameter for the net or gross value.

File: src/Core/Content/Product/Cart/ProductCartProcessor.php
Comment (OliverSkroblin) [inline src/Core/Content/Product/Cart/ProductCartProcessor.php]: I would not save the purchase price on the line item, because it is not relevant for the calculation and makes the line item itself unnecessarily large.

The purchase price should only be saved in the payload. In the payload you can save the gross and net price of the current currency price.

**Review by OliverSkroblin — CHANGES_REQUESTED** [review]

File: src/Core/Checkout/Cart/Rule/LineItemPurchasePriceGrossRule.php
Comment (hanneswernery) [reply src/Core/Checkout/Cart/Rule/LineItemPurchasePriceGrossRule.php]: That's a good suggestion. I added a small "select gross/net" field to the rule builder administration components. That way it is also easier to use for the administration user:
Screenshot 2020-04-02 at 09 49 48

**Comment by hanneswernery** [comment]
I implemented the requested changes. The changes are smaller now and everything seems to work fine 👍

File: src/Core/Content/Product/ProductEntity.php
Comment (ssltg) [inline src/Core/Content/Product/ProductEntity.php]: @deprecated tag:v6.3.0 use $purchasePrices instead 
ff

File: src/Core/Migration/Migration1582724349294AddNetAndGrossPurchasePrice.php
Comment (ssltg) [inline src/Core/Migration/Migration1582724349294AddNetAndGrossPurchasePrice.php]: Please test with 1 million Products and add a reasonable Limit and a for each to allow an update in chunks

**Review by jleifeld — APPROVED** [review]

File: src/Core/Framework/Demodata/Generator/ProductGenerator.php
Comment (ssltg) [inline src/Core/Framework/Demodata/Generator/ProductGenerator.php]: I would leave this out. So you will see if the Trigger Works fine.

File: src/Core/Framework/Test/DataAbstractionLayer/Write/TranslationTest.php
Comment (ssltg) [inline src/Core/Framework/Test/DataAbstractionLayer/Write/TranslationTest.php]: leave this out. 1. Checks if trigger works and 2. will not be forgotten when purchasePrice will be removed.

File: src/Core/Migration/Migration1582724349294AddNetAndGrossPurchasePrice.php
Comment (ssltg) [inline src/Core/Migration/Migration1582724349294AddNetAndGrossPurchasePrice.php:66]: The Trigger has to be added in a different way. We have a MR open for this currently. I will inform you when the MR is merged and synced and tell you what to change (its not much)

**Review by ssltg — CHANGES_REQUESTED** [review]
Looks good so far. A view little changes. The change with the commit has to wait until our change is merged. Hopefully this will be till tomorrow morning.

File: src/Core/Migration/Migration1582724349294AddNetAndGrossPurchasePrice.php
Comment (ssltg) [reply src/Core/Migration/Migration1582724349294AddNetAndGrossPurchasePrice.php:66]: In this MigrationFile you can see how the trigger needs to be set now

Comment (ssltg) [reply src/Core/Migration/Migration1582724349294AddNetAndGrossPurchasePrice.php:66]: src/Core/Migration/Migration1580218617RefactorShippingMethodPrice.php

**Comment by hanneswernery** [comment]
I updated the migration registration, rebased the branch onto the current master and squashed everything 👍

**Comment by ssltg** [comment]
We imported your PR and will merge it soon.
thanks for your contripution 👍

File: src/Administration/Resources/app/administration/src/module/sw-product/page/sw-product-detail/index.js
Comment (ssltg) [inline src/Administration/Resources/app/administration/src/module/sw-product/page/sw-product-detail/index.js:296]: Sadly we found a problem here.
You predefine the purchasePrices-Value with a priceStruct. This leads to a problem if the purchasePrice.net and .gross will be not set in the administration while creating a new product. When this product is pushed to the API, it will throw an error.
You predefined this to be able to use the price-component. We need to use a different attempt or a seprate component to be able to have either a priceStruct in product.purchasePrice or null.
currently, when purchasePrice is null on an exiting product, the purchasePrice fields are not shown.

Also there seems to be no way to set the purchasePrice in a different currency.

Comment (ssltg) [inline src/Administration/Resources/app/administration/src/module/sw-product/page/sw-product-detail/index.js:296]: Sadly we found a problem here.
You predefine the purchasePrices-Value with a priceStruct. This leads to a problem if the purchasePrice.net and .gross will be not set in the administration while creating a new product. When this product is pushed to the API, it will throw an error.
You predefined this to be able to use the price-component. We need to use a different attempt or a seprate component to be able to have either a priceStruct in product.purchasePrice or null.
currently, when purchasePrice is null on an exiting product, the purchasePrice fields are not shown.

Also there seems to be no way to set the purchasePrice in a different currency.

Comment (ssltg) [reply src/Administration/Resources/app/administration/src/module/sw-product/page/sw-product-detail/index.js:296]: Sorry for the late reply, the comment was stuck on github without being published

Comment (hanneswernery) [reply src/Administration/Resources/app/administration/src/module/sw-product/page/sw-product-detail/index.js:296]: What a coincidence, I just started working on this feature (again) to open a new PR for 6.2.3 😅  And I ran into the same wall here.

As I see it, nullable values are generally not supported in all input fields (e.g. `<sw-number-field>`), and quite impossible when also handling inheritance. I also asked about this in the Slack community channel (https://shopwarecommunity.slack.com/archives/C011VFQT7GB/p1593767643177500).

Without a major refactoring of the UI components, or the addition of a checkbox of some sort, I don't know how to add this nullable, inheritance-supporting purchase price field properly 😞 
Feel free to message me in the Slack community channel, if you have any ideas or thoughts on this.

**Comment by macidev** [comment]
Hi, are there any news regarding this feature? I'd like to have this feature in one of the next releases.

**Comment by fixpunkt** [comment]
Seems that this PR even made it to the roadmap at https://www.shopware.com/en/roadmap/:

image

Maybe @ssltg has an update on the status of the internal MR?

**Comment by bneumann97** [comment]
Hey @hanneswernery, thanks for your contribution 💙  Unfortunately, we have already done that. For this reason I'll close this PR.

**Comment by fixpunkt** [comment]
@bneumann97 would you mind linking the relevant commit? Did you end up implementing this based on this PR or was that a complete re-implementation?

**Comment by bneumann97** [comment]
@hanneswernery The commit was https://github.com/shopware/platform/commit/6943985ea1b5a88ab38f838b9914795416bb9a58 It looks like you are one of the authors of that commit, so I think it's based on your PR.

**Comment by mitelg** [comment]
Actually the changes came from this PR https://github.com/shopware/platform/pull/1155 :wink: which is an updated version of this PR. 

I don't know why @hanneswernery didn't closed this one :grin:

**Pull request closed without merging**
